fairscale/nn/data_parallel/fsdp_optim_utils.py

```diff
@@ -114,20 +131,39 @@ def _unflatten_optim_state(
     global_to_local_id = {}
     for local_id, num_unflat in enumerate(num_global_params):
         for _ in range(num_unflat):
-            global_to_local_id[next_global_id] = local_id
+            # Some params could be unused, which means the optimizer
+            # hasn't created their state. Therefore, `local_id` obtained
+            # by enumerating the params above could be out of the range
+            # of keys in `combined_state` above. Here is an example:
+            #
+            #    global    local    notes
+            #    0         0        FC1's weight, first flat buffer
+            #    1         0        FC1's bias, first flat buffer
+            #    2         None     FC2's weight, no flat state
+            #    3         None     FC2's bias, no flat state
+            #    4         2        FC3's weight, second flat buffer (but with id 2)
+            #    5         2        FC3's bias, second flat buffer (but with id 2)
+            global_to_local_id[next_global_id] = local_id if local_id in local_ids else None
             next_global_id += 1
     if not combined_state:
         return {}, global_to_local_id
 
     # copy non tensor state (like the "step" count) to all global entries
     unflat_state = {i: copy.deepcopy(non_tensor_state[0]) for i in range(sum(num_global_params))}
 
+    # remove the global entries that don't have optim state.
+    for g, l in global_to_local_id.items():
+        if l is None:
+            del unflat_state[g]
```

Comment:
Should we end up removing the entries that are unused in the global state dict? What are the cons of letting it remain in the dict without updating it?

Reply:
The optimizer only accepts parameters that are used and hence we want to be consistent with the state maintained there. Thats my reasoning for removing params entries but trying to understand if there is another.

Reply:
To remove them is to make it consistent with what pytorch optimizer does. The pytorch optimizer does not include "id:{}" for the unused params, so we don't include them here too. I will add a comment here to make it clear.